Выполнил задание 8-го спринта.

MaximusDenisov-YP opened the pull request.

Commit:
Добавил задачам время начала, конец и промежуток времени между ними.

- Изменил хранение задач в соответствии с изменениями для FileBackedTaskManager.java
- Исправил тесты.
- Добавлен метод для Epic, который высчитывает переменные начала выполнения, конца и промежутка времени между ними в зависимости от наполняемых его Subtasks.

Commit:
Добавлена логика проверки пересечений времени задач.

- Добавлен метод получения приоритезированного списка.
- Используемый и родительский класс Task теперь реализует интерфейс Comparable, что позволяет пользоваться им и его потомками взаимодействием с Tree-like коллекциями, также ввиду этого реализован метод compareTo() для Task и его потомков.
- Duration для Task like classes рассчитывается в минутах.
- Добавлен паттерн применения Option<>();

avfyodorov commented on src/main/ru/kanban/entity/Epic.java:
```diff
@@ -30,6 +46,31 @@ public Subtask putSubtaskToEpic(Subtask subTask) {
+                    .orElse(null)
+            );
+            if (this.getStartTime() != null && this.getEndTime() != null) {
+                this.setDuration(Duration.between(this.getStartTime(), this.getEndTime()));
```
По техническому заданию - "Продолжительность эпика — сумма продолжительностей всех его подзадач. "
А так считается разница между первой и последней.

avfyodorov commented on src/main/ru/kanban/manager/InMemoryTaskManager.java:
```diff
@@ -38,67 +38,50 @@ public ArrayList<Epic> getEpics() {
-            historyManager.add(result);
-        }
-        return result;
+    public TreeSet<Task> getPrioritizedTasks() {
```
К сожалению, данная реализация не соответствует требованиям ТЗ: 

.=============
Предполагается, что пользователь будет часто запрашивать этот список, поэтому подберите подходящую структуру данных для хранения. Сложность получения должна быть уменьшена ... до O(n)
...
Если сортировать список заново каждый раз, сложность получения будет O(n*log(n)). Можно хранить все задачи заранее отсортированными с помощью класса TreeSet.
.===========
Метод получения списка отсортированных задач должен выглядеть так: 
```java
	@Override
	public List<Task> getPrioritizedTasks() {
		return new ArrayList<>(prioritizedTasks);
	}
```
Нужно будет завести переменную - отсортированный список и поддерживать её в актуальном состоянии. 
То есть при каждом добавлении, изменении  задачи или подзадачи нужно будет проверять нет ли пересечения по времени с уже существующими в отсортированном списке. 
При удалении нужно будет находить в отсортированном списке соответствующие задачи и подзадачи и удалять их. 

````java 
protected final Set<Task> prioritizedTasks = new TreeSet<>(Comparator.comparing(Task::getStartTime));
```

avfyodorov commented on src/main/ru/kanban/manager/InMemoryTaskManager.java:
```diff
@@ -38,67 +38,50 @@ public ArrayList<Epic> getEpics() {
-        }
-        return result;
+    public TreeSet<Task> getPrioritizedTasks() {
+        TreeSet<Task> prioritizedTasks = new TreeSet<>();
```
Алгоритм проверки при добавлении (и изменении) задач и подзадач. 

Эпики в отсортированный список добавлять не нужно, потому что их параметры полностью определяются соответствующим подзадачами.

На всякий случай:

Если у задачи или подзадачи установлено время начала и продолжительность, но есть пересечение по времени, в этом случае не добавляем вообще никуда. 

Если время начала и продолжительность не установлены, то добавляем в общий список, а в отсортированный по времени нет, не добавляем. 

Если есть время начала и продолжительность и при этом нет пересечения по времени с другими задачами, то в этом случае добавляем и в общий список и в отсортированный по времени.

avfyodorov reviewed (commented):
Добрый день, Максим! 

В целом хорошо, правда, есть ряд уточнений.

Commit:
Произвёл исправления по замечаниям от ревьюера.

- Добавил/Исправил редактирование приоритезированного списка при добавлениях/изменениях/удалениях в задачах.
- Исправлено хранение приоритезированного списка.

avfyodorov reviewed (commented):
Добрый вечер, Максим!

Всё в порядке, замечаний нет. 
Работа принята.
